File: sdk/python/pyyol/skill/references/templates/mafia_agent.py
"""Mafia agent — 12 seats, hidden roles, phase machine.

Read references/games/mafia.md first. Note the view uses `legal`, NOT `legal_actions`,
and `day`/`phase` rather than `round`.

Replace `decide`; leave the rest.
"""

from __future__ import annotations

from typing import Tuple

from _shared import MatchMemory
from pyyol import Adapter
from pyyol.models import MafiaMove, MafiaView


class MafiaAgent(Adapter):
    name = "atlas-mafia"
    supported_games = ["mafia"]

    def __init__(self) -> None:
        self.mem = MatchMemory()

    def step(self, view: MafiaView) -> MafiaMove:
        # `legal` — not `legal_actions`. Getting this wrong means every move is
        # rejected and the engine plays for you.
        legal = view.legal or []
        if not legal:
            return MafiaMove(action="", rationale="nothing legal this phase")
        fallback = legal[0]

        # A turn is (day, phase) here — there is no round number.
        if self.mem.already_answered(view.match_id, (view.day, view.phase)):
            return MafiaMove(action=fallback, rationale="replayed turn")

        try:
            action, target, text, why = self.decide(view)
        except Exception as e:  # noqa: BLE001
            return MafiaMove(action=fallback, rationale=f"fallback: {e}")

        if action not in legal:
            action, target, text, why = fallback, -1, "", f"illegal action; {why}"

        # target stays -1 when unused: seat 0 is a REAL player, so a forgotten target
        # would otherwise silently act on them.
        return MafiaMove(
            action=action, target=target, text=text[:400], rationale=why[:200]
        )

    # --- your strategy -----------------------------------------------------

    def decide(self, view: MafiaView) -> Tuple[str, int, str, str]:
        """Return (action, target, text, reason).

        `view.public` is the table transcript; `view.private` carries what only you
        know (a Detective's finding arrives there and nowhere else). Rebuild your
        read of each seat from them every turn.
        """
        notes = self.mem.get(view.match_id)["notes"]
        living = [
            s for s, ok in (view.alive or {}).items() if ok and s != view.your_seat
        ]
        suspect = max(living, key=lambda s: notes.get(s, 0), default=-1)

        if "message" in (view.legal or []):
            return (
                "message",
                -1,
                f"Seat {suspect} has been quiet. Thoughts?",
                "opening a line on the current suspect",
            )
        if "vote" in (view.legal or []) and suspect >= 0:
            return "vote", suspect, "", f"voting {suspect}, my standing read"
        for act in ("investigate", "protect", "night_kill", "profile"):
            if act in (view.legal or []) and suspect >= 0:
                return act, suspect, "", f"{act} on {suspect}"
        return (view.legal or [""])[0], -1, "", "no better option this phase"


agent = MafiaAgent()

File: sdk/js/skill/references/templates/mafia_agent.mjs
/**
 * Mafia agent — social deduction, hidden roles, day/night phases.
 *
 * Read references/games/mafia.md first. Replace `decide`; leave the rest.
 */

import { Adapter } from "pyyol";

import { MatchMemory } from "./_shared.mjs";

export class MafiaAgent extends Adapter {
  name = "atlas-mafia";
  supportedGames = ["mafia"];

  #mem = new MatchMemory();

  step(view) {
    const legal = view.legal ?? [];
    if (legal.length === 0) return { action: "", rationale: "nothing legal this phase" };
    const fallback = legal.includes("pass") ? "pass" : legal[0];

    if (this.#mem.alreadyAnswered(view.match_id, [view.day, view.phase])) {
      return { action: fallback, rationale: "replayed turn" };
    }

    let action, target, text, why;
    try {
      ({ action, target = -1, text = "", why = "" } = this.decide(view));
    } catch (e) {
      return { action: fallback, rationale: `fallback: ${e instanceof Error ? e.message : e}` };
    }

    if (!legal.includes(action)) {
      [action, target, why] = [fallback, -1, `illegal action; ${why}`];
    }
    // NEVER target yourself, and never let a forgotten target become seat 0 — target defaults
    // to -1 ("no target") because seat 0 is a real player.
    if (target === view.your_seat) {
      target = -1;
      why = `refused to target myself; ${why}`;
    }

    return { action, target, text, rationale: String(why).slice(0, 200) };
  }

  // --- your strategy -------------------------------------------------------

  /**
   * Return { action, target, text, why }.
   *
   * `text` is your PUBLIC speech and rides along with the action — one model call produces both
   * the decision and what the table hears. `why` is private and never published, which matters
   * at night: publishing your reasoning would leak the mafia's plan to the town.
   */
  decide(view) {
    const legal = view.legal ?? [];
    const alive = Object.entries(view.alive ?? {})
      .filter(([seat, isAlive]) => isAlive && Number(seat) !== view.your_seat)
      .map(([seat]) => Number(seat));

    if (legal.includes("message")) {
      return { action: "message", text: "Watching who avoids committing.", why: "gathering reads" };
    }
    if (alive.length && (legal.includes("vote") || legal.includes("kill"))) {
      const action = legal.includes("vote") ? "vote" : "kill";
      return { action, target: alive[0], why: "no read yet; first living seat" };
    }
    return { action: legal[0], why: "first legal action" };
  }
}

export default MafiaAgent;
